core/libp2p/basic/reader.hpp

```diff
@@ -6,34 +6,31 @@
 #ifndef KAGOME_READER_HPP
 #define KAGOME_READER_HPP
 
-#include <vector>
-
+#include <boost/system/error_code.hpp>
 #include <gsl/span>
-#include <outcome/outcome.hpp>
 
 namespace libp2p::basic {
 
   struct Reader {
+    using ErrorCode = boost::system::error_code;
+    using ReadCallback = void(const ErrorCode & /*ec*/, size_t /*read bytes */);
+    using ReadCallbackFunc = std::function<ReadCallback>;
+
     virtual ~Reader() = default;
 
     /**
-     * @brief Blocks until all {@param bytes} bytes are read.
-     * @param bytes number of bytes to read
-     * @return vector of bytes or error code
+     * @brief Reads exactly {@code} out.size() {@nocode} bytes to the buffer.
+     * @param out output argument. Read data will be written to this buffer.
+     * @param cb callback with result of operation
      */
```

Comment:
Please, add an example of usage of those functions (writes also), so that the user understands, that he/she it to manage the read/write buffer, for example, by storing it in some external struct, which is not going to disappear after the read/write returns

Reply:
Of course, `span` already tells us it, but still, an example would be great

Reply:
Added comment